### PR TITLE
bugfix/issue-414-enum-unpacking

### PR DESCRIPTION
**Pull request by njlr**
Fix for https://github.com/fsprojects/FSharp.Data.GraphQL/issues/414

### CODE REVIEW

File: src/FSharp.Data.GraphQL.Server/Values.fs [hunk 1]
```diff
@@ -81,21 +97,29 @@ let rec internal compileByType (errMsg: string) (inputDef: InputDef): ExecuteInp
...
             | null -> none
             | coerced ->
                 let c = some coerced
-                if c <> null then c
-                else raise(GraphQLException (errMsg + notAssignableMsg innerdef coerced))
+
+                if c <> null then
+                    c
+                else
+                    raise (GraphQLException (errMsg + notAssignableMsg innerdef coerced))
```
Comment (xperiandri) [inline src/FSharp.Data.GraphQL.Server/Values.fs]:
```suggestion
                    raise <| GraphQLException (errMsg + notAssignableMsg innerdef coerced)
```

File: src/FSharp.Data.GraphQL.Server/Values.fs [hunk 2]
```diff
@@ -105,35 +129,57 @@ let rec internal compileByType (errMsg: string) (inputDef: InputDef): ExecuteInp
...
-        else none
+
+            if s <> null then
+                s
+            else
+                raise (
+                    GraphQLException
+                    <| errMsg
+                       + (sprintf "value of type %O is not assignable from %O" innerdef.Type (coerced.GetType ()))
+                )
```
Comment (xperiandri) [inline src/FSharp.Data.GraphQL.Server/Values.fs]:
```suggestion
                else raise <| GraphQLException ($"%s{errMsg}value of type %O{coerced.GetType()} is not assignable from %O")
```

File: src/FSharp.Data.GraphQL.Server/Values.fs [hunk 3]
```diff
@@ -105,35 +129,57 @@ let rec internal compileByType (errMsg: string) (inputDef: InputDef): ExecuteInp
...
+
         match input with
         | null when isNullable -> null
-        | null -> raise(GraphQLException <| errMsg + (sprintf "expected value of type %s, but no value was found." (vardef.TypeDef.ToString())))
+        | null ->
+            raise (
+                GraphQLException
+                <| errMsg
+                   + (sprintf "expected value of type %s, but no value was found." (vardef.TypeDef.ToString ()))
+            )
```
Comment (xperiandri) [inline src/FSharp.Data.GraphQL.Server/Values.fs]:
``` suggestion
        | null -> raise <| GraphQLException ($"%s{errMsg}expected value of type %s{vardef.TypeDef.ToString()}, but no value was found.")
```

File: src/FSharp.Data.GraphQL.Server/Values.fs [hunk 4]
```diff
@@ -148,23 +194,45 @@ let rec private coerceVariableValue isNullable typedef (vardef: VarDef) (input:
...
             mapped
-        | other -> raise (GraphQLException <| errMsg + (sprintf "Cannot coerce value of type '%O' to list." (other.GetType())))
-    | InputObject objdef ->
-        coerceVariableInputObject objdef vardef input (errMsg + (sprintf "in input object '%s': " objdef.Name))
+        | other ->
+            raise (
+                GraphQLException
+                <| errMsg
+                   + (sprintf "Cannot coerce value of type '%O' to list." (other.GetType ()))
+            )
```
Comment (xperiandri) [inline src/FSharp.Data.GraphQL.Server/Values.fs]:
``` suggestion
        | other -> raise <| GraphQLException ($"{errMsg}Cannot coerce value of type '%O{other.GetType()}' to list.")
```

File: src/FSharp.Data.GraphQL.Server/Values.fs [hunk 5]
```diff
@@ -148,23 +194,45 @@ let rec private coerceVariableValue isNullable typedef (vardef: VarDef) (input:
...
         | null when isNullable -> null
-        | null -> raise(GraphQLException <| errMsg + (sprintf "Expected Enum '%s', but no value was found." enumdef.Name))
-        | u when FSharpType.IsUnion(enumdef.Type) && enumdef.Type = input.GetType() -> u
-        | o when Enum.IsDefined(enumdef.Type, o) -> o
+        | null ->
+            raise (
+                GraphQLException
+                <| errMsg
+                   + (sprintf "Expected Enum '%s', but no value was found." enumdef.Name)
+            )
```
Comment (xperiandri) [inline src/FSharp.Data.GraphQL.Server/Values.fs]:
```suggestion
        | null -> raise <| GraphQLException ($"%s{errMsg}Expected Enum '%s{enumdef.Name}', but no value was found.")

```

File: src/FSharp.Data.GraphQL.Server/Values.fs [hunk 6]
```diff
@@ -148,23 +194,45 @@ let rec private coerceVariableValue isNullable typedef (vardef: VarDef) (input:
...
+            u
+        | o when Enum.IsDefined (enumdef.Type, o) -> o
         | _ ->
-            raise (GraphQLException <| errMsg + (sprintf "Cannot coerce value of type '%O' to type Enum '%s'" (input.GetType()) enumdef.Name))
-    | _ -> raise (GraphQLException <| errMsg + "Only Scalars, Nullables, Lists and InputObjects are valid type definitions.")
+            raise (
+                GraphQLException
+                <| errMsg
+                   + (sprintf "Cannot coerce value of type '%O' to type Enum '%s'" (input.GetType ()) enumdef.Name)
+            )
```
Comment (xperiandri) [inline src/FSharp.Data.GraphQL.Server/Values.fs]:
``` suggestion
         | _ -> raise <| GraphQLException ($"%s{errMsg}Cannot coerce value of type '%O{input.GetType()}' to type Enum '%s{enumdef.Name}'")

```

File: src/FSharp.Data.GraphQL.Server/Values.fs [hunk 7]
```diff
@@ -148,23 +194,45 @@ let rec private coerceVariableValue isNullable typedef (vardef: VarDef) (input:
...
+    | _ ->
+        raise (
+            GraphQLException
+            <| errMsg
+               + "Only Scalars, Nullables, Lists and InputObjects are valid type definitions."
+        )
```
Comment (xperiandri) [inline src/FSharp.Data.GraphQL.Server/Values.fs]:
```suggestion
    | _ -> raise <| GraphQLException ($"%s{errMsg}Only Scalars, Nullables, Lists, and InputObjects are valid type definitions.")
```

File: src/FSharp.Data.GraphQL.Server/Values.fs [hunk 8]
```diff
@@ -189,5 +259,5 @@ let internal coerceVariable (vardef: VarDef) (inputs) =
         | None ->
             match vardef.TypeDef with
             | Nullable _ -> null
-            | _ -> raise (GraphQLException (sprintf "Variable '$%s' of required type %s has no value provided." vname (vardef.TypeDef.ToString())))
+            | _ -> raise (GraphQLException (sprintf "Variable '$%s' of required type %s has no value provided." vname (vardef.TypeDef.ToString ())))
```
Comment (xperiandri) [inline src/FSharp.Data.GraphQL.Server/Values.fs]:
```suggestion
            | _ -> raise <| GraphQLException ($"Variable '$%s{vname}' of required type '%s{vardef.TypeDef.ToString ()}' has no value provided.")
```

**Review by xperiandri — CHANGES_REQUESTED** [review]
Could you highlight (comment) a piece of code that does a fix?
I cannot figure it out because of formatting

**Comment by njlr** [comment]
> Could you highlight (comment) a piece of code that does a fix? I cannot figure it out because of formatting

Sorry, I ran Fantomas since I thought that was the policy going forward. 

You can see the logic changes in the first commit: https://github.com/fsprojects/FSharp.Data.GraphQL/pull/415/commits/e4b6e90e6e5183ac22cc838a0368b6ab4b09fee3

**Comment by xperiandri** [comment]
Thanks. Looks good. Let's reformat exceptions raising and we a good to merge

**Comment by njlr** [comment]
> Thanks. Looks good. Let's reformat exceptions raising and we a good to merge

Addressed. Please squash this on merging - GitHub's "commit suggestion" feature is not ideal

BTW, is there a plan for when the next release of this library will land?

File: src/FSharp.Data.GraphQL.Server/Values.fs [hunk 9]
```diff
@@ -105,35 +130,51 @@ let rec internal compileByType (errMsg: string) (inputDef: InputDef): ExecuteInp
...
     match typedef with
     | Scalar scalardef ->
         match scalardef.CoerceValue input with
         | None when isNullable -> null
         | None ->
-            raise (GraphQLException <| errMsg + (sprintf "expected value of type %s but got None" scalardef.Name))
+            raise (
+                GraphQLException
+                <| errMsg
+                   + (sprintf "expected value of type %s but got None" scalardef.Name)
```
Comment (xperiandri) [inline src/FSharp.Data.GraphQL.Server/Values.fs]:
```suggestion
                <| $"%s{errMsg}expected value of type %s{scalardef.Name} but got None")
```

File: src/FSharp.Data.GraphQL.Server/Values.fs [hunk 10]
```diff
@@ -148,23 +189,37 @@ let rec private coerceVariableValue isNullable typedef (vardef: VarDef) (input:
...
+            ->
+            u
+        | o when Enum.IsDefined (enumdef.Type, o) -> o
         | _ ->
-            raise (GraphQLException <| errMsg + (sprintf "Cannot coerce value of type '%O' to type Enum '%s'" (input.GetType()) enumdef.Name))
-    | _ -> raise (GraphQLException <| errMsg + "Only Scalars, Nullables, Lists and InputObjects are valid type definitions.")
+            raise (
+                GraphQLException
+                <| errMsg
+                   + (sprintf "Cannot coerce value of type '%O' to type Enum '%s'" (input.GetType ()) enumdef.Name)
```
Comment (xperiandri) [inline src/FSharp.Data.GraphQL.Server/Values.fs]:
```suggestion
                <| $"%s{errMsg}Cannot coerce value of type '%O{input.GetType ()}' to type Enum '%s{enumdef.Name}'")
```

**Review by xperiandri — CHANGES_REQUESTED** [review]
A few more strings

**Comment by xperiandri** [comment]
@njlr yes I'll squash
@Juriyx almost finished a pipeline. So you will be able to consume these fixes at least from the GitHub feed the next week.
Release of 2.0 I suppose till the end of the year or January 2023

**Comment by njlr** [comment]
> @njlr yes I'll squash @Juriyx almost finished a pipeline. So you will be able to consume these fixes at least from the GitHub feed the next week. Release of 2.0 I suppose till the end of the year or January 2023

All done. 

Looking forward to the improvements! :rocket:

File: src/FSharp.Data.GraphQL.Server/Values.fs [hunk 11]
```diff
@@ -148,23 +188,36 @@ let rec private coerceVariableValue isNullable typedef (vardef: VarDef) (input:
...
                 |> List.rev
                 |> List.fold (fun acc coerced -> cons coerced acc) nil
+
             mapped
-        | other -> raise (GraphQLException <| errMsg + (sprintf "Cannot coerce value of type '%O' to list." (other.GetType())))
-    | InputObject objdef ->
-        coerceVariableInputObject objdef vardef input (errMsg + (sprintf "in input object '%s': " objdef.Name))
+        | other ->
+            raise
+            <| GraphQLException ($"{errMsg}Cannot coerce value of type '%O{other.GetType ()}' to list.")
```
Comment (xperiandri) [inline src/FSharp.Data.GraphQL.Server/Values.fs:195]:
```suggestion
            <| GraphQLException ($"%s{errMsg}Cannot coerce value of type '%O{other.GetType ()}' to list.")
```

File: src/FSharp.Data.GraphQL.Server/Values.fs [hunk 12]
```diff
@@ -105,35 +130,50 @@ let rec internal compileByType (errMsg: string) (inputDef: InputDef): ExecuteInp
...
+            none
     | List (Input innerdef) ->
         let cons, nil = ReflectionHelper.listOfType innerdef.Type
+
         match input with
         | null when isNullable -> null
-        | null -> raise(GraphQLException <| errMsg + (sprintf "expected value of type %s, but no value was found." (vardef.TypeDef.ToString())))
+        | null ->
+            raise
+            <| GraphQLException ($"%s{errMsg}expected value of type %s{vardef.TypeDef.ToString ()}, but no value was found.")
```
Comment (xperiandri) [inline src/FSharp.Data.GraphQL.Server/Values.fs]:
```suggestion
            <| GraphQLException ($"%s{errMsg}expected value of type '%s{vardef.TypeDef.ToString ()}', but no value was found.")
```

File: src/FSharp.Data.GraphQL.Server/Values.fs [hunk 13]
```diff
@@ -105,35 +130,50 @@ let rec internal compileByType (errMsg: string) (inputDef: InputDef): ExecuteInp
...
-            if s <> null
-            then s
-            else raise (GraphQLException <| errMsg + (sprintf "value of type %O is not assignable from %O" innerdef.Type (coerced.GetType())))
-        else none
+
+            if s <> null then
+                s
+            else
+                raise
+                <| GraphQLException ($"%s{errMsg}value of type %O{innerdef.Type} is not assignable from %O{coerced.GetType ()}")
```
Comment (xperiandri) [inline src/FSharp.Data.GraphQL.Server/Values.fs]:
```suggestion
                <| GraphQLException ($"%s{errMsg}value of type '%O{innerdef.Type}' is not assignable from '%O{coerced.GetType ()}'.")
```

File: src/FSharp.Data.GraphQL.Server/Values.fs [hunk 14]
```diff
@@ -105,35 +130,50 @@ let rec internal compileByType (errMsg: string) (inputDef: InputDef): ExecuteInp
...
+let rec private coerceVariableValue isNullable typedef (vardef : VarDef) (input : obj) (errMsg : string) : obj =
     match typedef with
     | Scalar scalardef ->
         match scalardef.CoerceValue input with
         | None when isNullable -> null
         | None ->
-            raise (GraphQLException <| errMsg + (sprintf "expected value of type %s but got None" scalardef.Name))
+            raise (
+                GraphQLException
+                <| $"%s{errMsg}expected value of type %s{scalardef.Name} but got None"
```
Comment (xperiandri) [inline src/FSharp.Data.GraphQL.Server/Values.fs]:
```suggestion
                <| $"%s{errMsg}expected value of type '%s{scalardef.Name}' but got 'None'."
```

File: src/FSharp.Data.GraphQL.Server/Values.fs [hunk 15]
```diff
@@ -148,23 +188,36 @@ let rec private coerceVariableValue isNullable typedef (vardef: VarDef) (input:
...
+            && enumdef.Type = input.GetType ()
+            ->
+            u
+        | o when Enum.IsDefined (enumdef.Type, o) -> o
         | _ ->
-            raise (GraphQLException <| errMsg + (sprintf "Cannot coerce value of type '%O' to type Enum '%s'" (input.GetType()) enumdef.Name))
-    | _ -> raise (GraphQLException <| errMsg + "Only Scalars, Nullables, Lists and InputObjects are valid type definitions.")
+            raise (
+                GraphQLException
+                <| $"%s{errMsg}Cannot coerce value of type '%O{input.GetType ()}' to type Enum '%s{enumdef.Name}'"
```
Comment (xperiandri) [inline src/FSharp.Data.GraphQL.Server/Values.fs]:
```suggestion
                <| $"%s{errMsg}Cannot coerce value of type '%O{input.GetType ()}' to type Enum '%s{enumdef.Name}'."
```

**Review by xperiandri — CHANGES_REQUESTED** [review]
Looks like we had this inconsistency in messages. Are all of them good with these changes?

**Comment by xperiandri** [comment]
I already have a local commit with fix if you agree with my comments

**Comment by njlr** [comment]
> I already have a local commit with fix if you agree with my comments

Please go ahead!

**Pull request merged**